[SelectionDAG] Add support for the 3-way comparison intrinsics [US]CMP

Poseydon42 opened the pull request:
This PR adds initial support for the `scmp`/`ucmp` 3-way comparison intrinsics in the SelectionDAG.

What works as of now:
* An invokation of the intrinsic in the IR gets properly lowered into SelectionDAG
* A node with opcodes `UCMP`/`SCMP` gets properly expanded into two comparisons and two selects
* Narrow scalar arguments and return types are properly handled (i.e. i3 or i51)
* Wide scalar arguments and return types are properly handled (i.e. i87 or i139)
* Vector arguments and return types are properly widened/split where necessary

Poseydon42 committed:
[SelectionDAG] Add support for the 3-way comparison intrinsics [US]CMP

nikic commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -677,6 +677,11 @@ enum NodeType {
   UMAX,
 
+  /// [US]CMP - 3-way comparison of signed or unsigned integers. Returns -1, 0,
+  /// or 1 depending on whether Op0 <, ==, or > Op1
```
I'd mention that the result and operand types can differ here.

nikic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -1246,6 +1251,14 @@ SDValue DAGTypeLegalizer::PromoteIntRes_SADDSUBO(SDNode *N, unsigned ResNo) {
+  EVT PromotedResultTy =
+      TLI.getTypeToTransformTo(*DAG.getContext(), N->getValueType(0));
+  SDValue Result = DAG.getNode(N->getOpcode(), SDLoc(N), PromotedResultTy,
+                               N->getOperand(0), N->getOperand(1));
```
Can directly return here, no need to store in a variable first.

nikic commented on llvm/test/CodeGen/X86/uscmp.ll:
```diff
@@ -0,0 +1,163 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -show-mc-encoding | FileCheck %s --check-prefixes=CHECK,NO-NDD
```
```suggestion
; RUN: llc < %s -mtriple=x86_64-unknown-unknown | FileCheck %s
```

No need to check encoding in this test. You'll probably have to manually drop the NDD lines.

Poseydon42 committed:
[SelectionDAG] Properly handle legalization of vector types for [US]CMP nodes

nikic reviewed (commented):
`build/bin/llc -mtriple=aarch64-unknown-linux-gnu < llvm/test/CodeGen/X86/uscmp.ll` gives a selection error.

nikic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -1327,6 +1344,31 @@ void DAGTypeLegalizer::SplitVecRes_TernaryOp(SDNode *N, SDValue &Lo,
+                              DAG.getVectorIdxConstant(0, dl));
+  SDValue LHSHi = DAG.getNode(
+      ISD::EXTRACT_SUBVECTOR, dl, SplitOpVT, LHS,
+      DAG.getVectorIdxConstant(SplitOpVT.getVectorNumElements(), dl));
```
I think you're looking for `DAG.SplitVector()` here.

Poseydon42 replied:
Thank you, didn't know it existed :)

nikic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -6273,6 +6382,32 @@ SDValue DAGTypeLegalizer::WidenVecOp_EXTEND(SDNode *N) {
+  ISD::NodeType ExtendOpcode =
+      N->getOpcode() == ISD::SCMP ? ISD::SIGN_EXTEND : ISD::ZERO_EXTEND;
+  LHS = DAG.getNode(ExtendOpcode, dl, ResVT, LHS);
+  RHS = DAG.getNode(ExtendOpcode, dl, ResVT, RHS);
```
I'm confused by the WidenVecOp/WidenVecRes implementations. The widening here is in the sense of increasing the number of vector elements, not the size of the element type, so why are extends relevant here?

Poseydon42 replied:
I may be very mistaken here, but as far as I understand the other option would have been to use widened operands (i.e. `v4i8` widened to `v16i8`). This would in turn cause the result type to be widened as well (i.e. in the same scenario `v4i32` might have had to be widened to `v16i32`). The result type might now become too wide, requiring it to be split, which would split the operands, creating a legalization cycle and/or an assertion failure down the line (when I tried to remove sign extension just now I triggered an assertion because a mask and operands of `VSELECT` have different number of arguments, but I believe that I had also run into the described legalization cycle before).

If we just extract the subvector from a widened operand without sign-extending it, we wouldn't actually widen the operand at all, i.e. the sequence of type changes might look like `v4i8` (original type) -> `v16i8` (widened type) -> `v4i8` (type extracted from the widened vector), which would mean that we didn't change the instruction at all and instead just ran into another cycle.

What I have decided to do instead is after we get the widened operand and extract the necessary number of elements from it, we can then sign extend this extracted vector to the type of the result of the instruction. We know that this type is legal at this stage and that the number of elements in the result and the operands must match, so the sign extension is technically valid and we have shifted the job of legalizing a vector operand with a small element size to `SIGN_EXTEND`/`ZERO_EXTEND`. Whether this is a good idea or not I am not sure and would be happy to hear your opinion on it.

Poseydon42 commented on llvm/test/CodeGen/X86/uscmp.ll:
```diff
@@ -0,0 +1,498 @@
+; CHECK-NEXT:    cmovael %ecx, %eax
+; CHECK-NEXT:    # kill: def $al killed $al killed $eax
+; CHECK-NEXT:    retq
+  %1 = call i8 @llvm.ucmp(i32 %x, i32 %y)
```
Did just that :)

dtcxzyw commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7143,6 +7143,22 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+  case Intrinsic::scmp: {
+    SDValue Op1 = getValue(I.getArgOperand(0));
+    SDValue Op2 = getValue(I.getArgOperand(1));
+    EVT DestVT = DAG.getTargetLoweringInfo().getValueType(DAG.getDataLayout(),
```
```suggestion
    EVT DestVT = TLI.getValueType(DAG.getDataLayout(),
```

See https://github.com/llvm/llvm-project/blob/4b5e0a1d12f0d624852c6d7c82207647ed5282aa/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L6342-L6346

dtcxzyw commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -164,6 +164,12 @@ void DAGTypeLegalizer::ScalarizeVectorResult(SDNode *N, unsigned ResNo) {
+
+  case ISD::SCMP:
+  case ISD::UCMP:
+    R = ScalarizeVecRes_CMP(N);
```
Missing test for scmp + scalarizevec

Poseydon42 replied:
Added one

nikic commented:
```llvm
define <2 x i16> @test_ucmp.2.8.16(<2 x i8> %x, <2 x i8> %y) {
  %1 = call <2 x i16> @llvm.ucmp(<2 x i8> %x, <2 x i8> %y)
  ret <2 x i16> %1
}
```
Asserts:
```
llc: /home/npopov/repos/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:7434: SDValue llvm::SelectionDAG::getNode(unsigned int, const SDLoc &, EVT, SDValue, SDValue, SDValue, const SDNodeFlags): Assertion `(VT.isScalableVector() != N2VT.isScalableVector() || VT.getVectorMinNumElements() >= N2VT.getVectorMinNumElements()) && "Insert subvector must be from smaller vector to larger vector!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build/bin/llc test.ll
1.	Running pass 'Function Pass Manager' on module 'test.ll'.
2.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@test_ucmp.2.8.16'
 #0 0x0000000006474828 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build/bin/llc+0x6474828)
 #1 0x00000000064723ee llvm::sys::RunSignalHandlers() (build/bin/llc+0x64723ee)
 #2 0x0000000006474ed8 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f2303c53710 __restore_rt (/lib64/libc.so.6+0x40710)
 #4 0x00007f2303cab144 __pthread_kill_implementation (/lib64/libc.so.6+0x98144)
 #5 0x00007f2303c5365e gsignal (/lib64/libc.so.6+0x4065e)
 #6 0x00007f2303c3b902 abort (/lib64/libc.so.6+0x28902)
 #7 0x00007f2303c3b81e _nl_load_domain.cold (/lib64/libc.so.6+0x2881e)
 #8 0x00007f2303c4b977 (/lib64/libc.so.6+0x38977)
 #9 0x0000000006249fd7 llvm::SelectionDAG::getNode(unsigned int, llvm::SDLoc const&, llvm::EVT, llvm::SDValue, llvm::SDValue, llvm::SDValue, llvm::SDNodeFlags) (build/bin/llc+0x6249fd7)
#10 0x00000000062f4a41 llvm::DAGTypeLegalizer::WidenVecRes_CMP(llvm::SDNode*) LegalizeVectorTypes.cpp:0:0
#11 0x00000000062ed2bd llvm::DAGTypeLegalizer::WidenVectorResult(llvm::SDNode*, unsigned int) LegalizeVectorTypes.cpp:0:0
#12 0x0000000006294fdb llvm::DAGTypeLegalizer::run() LegalizeTypes.cpp:0:0
```

I'm using this script to generate test cases:
```php
<?php

function makeType(int $numElems, int $bw): string {
  $ty = 'i' . $bw;
  if ($numElems != 0) {
    return '<' . $numElems . ' x ' . $ty . '>';
  }
  return $ty;
}

$bws = [1, 2, 3, 4, 7, 8, 9, 15, 16, 17, 31, 32, 33, 63, 64, 65, 127, 128, 129, 255, 256, 257];

foreach (['ucmp', 'scmp'] as $intrin) {
  for ($numElems = 0; $numElems < 100; ++$numElems) {
    foreach ($bws as $argBW) {
      foreach ($bws as $resBW) {
        if ($resBW == 1) {
          continue;
        }
        $argType = makeType($numElems, $argBW);
        $resType = makeType($numElems, $resBW);
        echo "define $resType @test_$intrin.$numElems.$argBW.$resBW($argType %x, $argType %y) {\n";
        echo "  %1 = call $resType @llvm.$intrin($argType %x, $argType %y)\n";
        echo "  ret $resType %1\n";
        echo "}\n";
        echo "\n";
      }
    }
  }
}
```

nikic commented on llvm/test/CodeGen/X86/uscmp.ll:
```diff
@@ -0,0 +1,498 @@
+  ret <16 x i8> %1
+}
+
+define <7 x i25> @scmp_uncommon_vectors(<7 x i109> %x, <7 x i109> %y) {
```
This comment is still open.

Poseydon42 replied:
Added it to every test case just to make sure :)

Poseydon42 commented:
Thank you for providing the script, I've tested the current revision of the code against the generated tests and it seems to work on both x86_64 and AArch64. I've reduced the maximum size of vector types from 100 to 50 as it was taking ages otherwise, but I don't think it should change anything.

nikic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -1327,6 +1368,21 @@ void DAGTypeLegalizer::SplitVecRes_TernaryOp(SDNode *N, SDValue &Lo,
+  EVT SplitResVT = N->getValueType(0).getHalfNumVectorElementsVT(Ctxt);
+
+  auto [LHSLo, LHSHi] = DAG.SplitVector(LHS, dl);
+  auto [RHSLo, RHSHi] = DAG.SplitVector(RHS, dl);
```
Might make sense to add a check for whether the value type also has TypeSplitVector action and call GetSplitVector in that case? Similar to what you do in ScalarizeVecRes_CMP.

Poseydon42 replied:
Fixed

nikic commented on llvm/test/CodeGen/X86/ucmp.ll:
```diff
@@ -0,0 +1,972 @@
+; CHECK-NEXT:    retq
+  %1 = call i8 @llvm.ucmp(i32 %x, i32 %y)
+  ret i8 %1
+}
```
I'd also add some tests for the practically relevant cases, in addition to the legalization coverage. cmp i32 with i8 result is one of them, but also cmp i128, i64, i16 and i8 with i8 result. And then maybe also i32/i64 result type.

Those are the ones where we would care about good codegen, the rest mostly just needs to work...

nikic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4426,6 +4511,56 @@ SDValue DAGTypeLegalizer::WidenVecRes_Binary(SDNode *N) {
+  ElementCount WidenResEC = WidenResVT.getVectorElementCount();
+  EVT WidenResElementVT = WidenResVT.getVectorElementType();
+
+  // At this point we know that the type of operands should be legal,
```
This doesn't sound right. The result types are legalized first, so we don't know at this point that the operands are legal.

Poseydon42 replied:
Perhaps I've worded this comment a bit poorly. What I meant is that since (unless I'm massively misunderstanding something) the node that produces a value is legalized strictly before any nodes that use this value, the actual `SDValue`s of the operands that we retrieve a few lines above this comment are guaranteed to be correctly widened. I am not 100% confident that those types would be fully legal (i.e. the type wouldn't be something like <4 x i29>), but it shouldn't matter in this scenario I believe since `WidenVecRes_CMP` makes sure that the CMP node is retyped so that the return type and operand types match, so even if any further legalization is required after this point, the fact that the result and operand types are the same prevents us from running into legalization cycles. Hope this clarifies things a bit, and I will rewrite that comment a bit better shortly.

nikic replied:
> Perhaps I've worded this comment a bit poorly. What I meant is that since (unless I'm massively misunderstanding something) the node that produces a value is legalized strictly before any nodes that use this value, the actual SDValues of the operands that we retrieve a few lines above this comment are guaranteed to be correctly widened.

The legalization happens from the root of the SDAG to the leaves. This means that nodes using a value are legalized *before* nodes producing the value.

I think your general point still stands because you explicitly look up the type the operands will get widened to.

Poseydon42 replied:
> This means that nodes using a value are legalized before nodes producing the value.

I don't think I agree with this -- as far as I'm aware it's the direct opposite, a node can only be legalized after all of its operands are legalized. That's why when you run `llc` with debug loggging enabled the first nodes to be legalized are things like target registers and constants, and the last node is always return. This is what allows me to use the approach I'm using, as I know that inside `WidenVecRes_CMP` the two operands have already been widened, but the CMP node itself still points to the old, unwidened operands.

If you also look into the implementation of `DAGTypeLegalizer::run`, the one of the first things it does is populate the worklist of nodes to be processed with those that have no operands, and then updates the worklist any time all arguments of some node are processed.

nikic replied:
Yeah, you're right. Ignore me.

nikic reviewed (commented):
This looks mostly good to me -- main uncertainty is around the vector widening implementations, which are pretty unusual. It would be good if an SDAG expert could look over that part at least (@RKSimon maybe?)

nikic commented:
It would also be good to add some test coverage for other targets -- for those cases, I don't think we need the full legalization test coverage, but mainly some scalar tests with common types, so we can see how the codegen looks like and have a baseline when improving it later. (The currently used expansion is probably not the best default expansion, but it will be easier to judge that with baseline test coverage.)

Poseydon42 commented:
Added the tests that were suggested (for AArch64 and x86 only for now, but if any other architectures are suggested I'd be happy to add those too). Also fixed the slightly misleading comment in `WidenVecRes_CMP` and addressed the comment about checking whether operands should also be split in `SplitVecRes_CMP`.

RKSimon commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -677,6 +677,12 @@ enum NodeType {
 
+  /// [US]CMP - 3-way comparison of signed or unsigned integers. Returns -1, 0,
+  /// or 1 depending on whether Op0 <, ==, or > Op1. The operands can have type
+  /// different to the result.
```
Should we enforce that for vector types the result type should match the operand type? Treating these as close to regular (non-commutable) binops as possible makes sense to me.

Poseydon42 replied:
I feel like allowing them to be different isn't that much of an issue, while potentially allowing us to emit more optimal code (maybe?). After all, if you're comparing two vectors of `i128`s you don't need the result to be a vector of i128 as well, and I'm not sure whether forcing the return type to be the same as operands in this scenario would be handled optimally by other optimizations in SDAG pipeline.

RKSimon commented on llvm/test/CodeGen/X86/scmp.ll:
```diff
@@ -0,0 +1,985 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown | FileCheck %s
```
Please can you add i686-- test coverage as well

Poseydon42 replied:
Done

Poseydon42 committed:
Added tests for both x86 and x86_64

nikic commented:
This one asserts with `-mattr=+avx512f`:
```llvm
define <33 x i2> @test_ucmp.33.4.2(<33 x i4> %x, <33 x i4> %y) {
  %1 = call <33 x i2> @llvm.ucmp(<33 x i4> %x, <33 x i4> %y)
  ret <33 x i2> %1
}
```

Poseydon42 committed:
Changed WidenVecRes_CMP to unroll the operation to avoid crashing with AVX-512

Poseydon42 commented:
Fixed the crash with AVX-512. `WidenVecRes_CMP` now unrolls the operation into a bunch of scalar operations and then rebuilds a vector whenever the widened operand and result have a different number of elements. There is probably a way to avoid doing this in every situation, but I think it's worth adding that later and in a separate PR.

nikic reviewed (approved):
LGTM
